libbeat/processors/dissect/field.go

```diff
@@ -299,5 +397,13 @@ func extractKeyParts(rawKey string) (key string, ordinal int, length int, greedy
 		greedy = true
 	}
 
-	return m[0][1], ordinal, length, greedy
+	if m[0][7] != "" {
+		if m[0][8] == "" {
+			dataType = unsetDataType
```

Comment:
Nit. Nothing serious, but I am a bit concerned about this `unsetDataType`. There are two cases here we we might say that we have an "unset" datatype, one when the pipe (`|`) is not used, then the expression is correct but there is no datatype set, the other one when the pipe is used, but datatype is empty, what is incorrect. I think this may cause confusion if future readings of the code.
I wonder if we could directly expect in the regular expression that the datatype is not empty, and then check the validity of the resulting fields in a more similar way to how other things like conflicting fields are checked [here](https://github.com/elastic/beats/pull/18683/files#diff-4376db67dcdbf120310a4ec6e1523f4dL209).
Or maybe another option is to make the `dataType` a pointer to string, so we use `nil` when the datatype is really not set, but the expression is correct (no `|` is used), and the empty string when `|` is used but is empty.

Just some thoughts, no need to change it, as you prefer :slightly_smiling_face:

Reply:
Thanks for the suggestion, refactored code and added check for suffix |, when data type is missing. Removed need for unset to check for missing data type.

Reply:
Thanks for this refactor!